refactoring voter roll and delete election controllers

mikefranze opened the pull request:
First pass at refactoring the voter roll controllers and the delete election controllers. They didn't require multiple databases, but I have a few TODOs for after Scott's PR is merged. I combined a few of them in the same file where it looks like it made sense.

ArendPeter commented:
I think the elections controller still needs to be deleted?

mikefranze commented:
> I think the elections controller still needs to be deleted?

Thanks, old middleware deleted.

ArendPeter commented on backend/src/Controllers/voterRolls.controller.ts:
```diff
@@ -30,11 +30,6 @@ const getRollsByElectionID = async (req: any, res: any, next: any) => {
-    res.json({ election: req.election, electionRoll: req.electionRoll });
-}
-
 const addElectionRoll = async (req: any, res: any, next: any) => {
```
Should this snippet be deleted since it was replace by addElectionRollController.ts?

mikefranze replied:
That is still used in the create election route which scott is working on.

ArendPeter commented on backend/src/Controllers/getElectionRollController.ts:
```diff
@@ -0,0 +1,65 @@
+
+const className = "VoterRolls.Controllers";
+
+const getRollsByElectionID = async (req: any, res: any, next: any) => {
```
I think the corresponding function in backend/src/Controllers/voterRolls.controller.ts needs to be deleted?

mikefranze replied:
That is still used by the '/Election/:id/finalize' route which gets the rolls and sends the invitations.

ArendPeter reviewed (changes requested):
Thanks for adding the updates! I think there's still a few missing though

ArendPeter reviewed (approved):
Looks good, thanks for the clarifications!

scottPlusPlus commented on backend/src/Routes/elections.routes.js:
```diff
@@ -6,65 +6,42 @@ const electionController = require('../Controllers/elections.controllers')
+router.post('/Election/:id/rolls/flag', authController.getUser, changeElectionRollStateController.flagElectionRoll)
+router.post('/Election/:id/rolls/invalidate', authController.getUser, changeElectionRollStateController.invalidateElectionRoll)
+router.post('/Election/:id/rolls/unflag', authController.getUser, changeElectionRollStateController.uninvalidateElectionRoll)
+router.post('/Election/:id/rolls/', authController.getUser, addElectionRoll)
```
In general, I'd like to see the controllers handle everything they need explicitly, rather than implicitly assuming that other methods have been chained in the route that stash data into the req.

scottPlusPlus replied:
If it's really every route, then yes exactly we can use a middleware  (like you did for the catchError), and then add the user data explicitly to the IRequest (even if it's nullable).

If it's going to be conditional per route, I'd suggest putting it in a single function to handle the route, so that all the code required for the route is in one place.

mikefranze commented:
- Added async handler to the routes that will catch any errors and throw them to our custom error handler middleware. 
- Moved getUser to main app middleware and moved user_auth logic to getElectionByID middleware that runs whenever we have the id route param. 
- Added expect permission controller util, we can start to phase out where we use hasPermission functions.

scottPlusPlus commented:
LGTM!
